h/views/badge.py

```diff
@@ -22,6 +24,17 @@ def record_metrics(count,
     record_metric('Custom/Badge/badgeCountIsZero', int(count == 0))
 
 
+def _has_uri_ever_been_annotated(db, uri):
+    """Return `True` if a given URI has ever been annotated."""
+
+    # This check is written with SQL directly to guarantee an efficient
+    # query and minimize SQLAlchemy overhead.
+    db_conn = db.connection()
+    query = text('SELECT EXISTS(SELECT 1 FROM document_uri WHERE uri_normalized = :uri)')
```

Comment:
I guess we're using `document_uri`, rather than `annotation`, because `document_uri` has less rows. Nice

Reply:
The main reason is that we have an existing index on `document_uri.uri_normalized` so this query can be answered efficiently. We don't appear to have an index on `annotation.target_uri_normalized`.